### ISSUE TITLE
Large CPU/GPU numerical discrepancy in tf.linalg.matmul for float32 linear layer equivalent

### ISSUE THREAD

**Issue by dingjiawei65** [tensorflow/tensorflow#127346]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

TensorFlow 2.20.0

### Custom code

Yes

### OS platform and distribution

 Ubuntu 22.04.5 LTS, x86_64

### Python version

Python 3.11.14

### Current behavior?

CPU and GPU outputs for the same float32 inputs should be reasonably close, or the expected tolerance / numerical behavior should be documented clearly.
In this case, the observed max absolute difference is much larger than a typical small numerical tolerance such as 1e-3。

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf

for gpu in tf.config.list_physical_devices("GPU"):
    tf.config.experimental.set_memory_growth(gpu, True)

rng = np.random.default_rng(133006)

x_np = (rng.standard_normal((64, 197, 1536)).astype("float32") * 1000.0)
w_np = (rng.standard_normal((384, 1536)).astype("float32") * 50.0)
b_np = (rng.standard_normal((384,)).astype("float32") * 50.0)

with tf.device("/CPU:0"):
    x = tf.constant(x_np)
    w = tf.constant(w_np)
    b = tf.constant(b_np)
    dropped_cpu = tf.keras.layers.Dropout(rate=0.0)(x, training=False)
    y_cpu = tf.linalg.matmul(dropped_cpu, w, transpose_b=True) + b
    y_cpu = y_cpu.numpy()

with tf.device("/GPU:0"):
    x = tf.constant(x_np)
    w = tf.constant(w_np)
    b = tf.constant(b_np)
    dropped_gpu = tf.keras.layers.Dropout(rate=0.0)(x, training=False)
    y_gpu = tf.linalg.matmul(dropped_gpu, w, transpose_b=True) + b
    y_gpu = y_gpu.numpy()

dropout_diff = np.max(np.abs(dropped_cpu.numpy() - dropped_gpu.numpy()))
linear_diff = np.max(np.abs(y_cpu - y_gpu))
mean_diff = np.mean(np.abs(y_cpu - y_gpu))
max_output = np.max(np.abs(y_cpu))

print("TensorFlow:", tf.__version__)
print("GPUs:", tf.config.list_physical_devices("GPU"))
print("output shape:", y_cpu.shape)
print("dropout CPU/GPU max abs diff:", dropout_diff)
print("linear CPU/GPU max abs diff:", linear_diff)
print("linear CPU/GPU mean abs diff:", mean_diff)
print("max abs output:", max_output)
print("relative max diff:", linear_diff / max_output)
```

### Relevant log output

```shell
TensorFlow: 2.20.0
GPUs: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU')]
output shape: (64, 197, 384)
dropout CPU/GPU max abs diff: 0.0
linear CPU/GPU max abs diff: 3066.0
linear CPU/GPU mean abs diff: 459.5823
max abs output: 10401001.0
relative max diff: 0.00029477
```

**Comment by Kayyuri** [comment]
Hi @dingjiawei65, Thank you for reporting this issue!

You've clearly demonstrated the numerical discrepancies in `tf.linalg.matmul` for large-scale `float32` inputs. Your detailed description, clear reproduction steps, and the provided code snippet are very helpful. We will investigate this from our side and provide an update.

**Comment by RiccardoSquarcialupi** [comment]
Hey @dingjiawei65, I think this might just be TF32 rather than an actual bug.

Ampere+ GPUs run float32 matmul through TensorFloat-32 by default, which only keeps ~10 bits of mantissa instead of float32's 23 — so some precision loss vs CPU is expected, and it gets more noticeable with a big contracted dimension like your 1536.

I tried simulating TF32-style rounding on your exact inputs and got a relative diff in the same ballpark as what you saw (2.9477e-4), while plain float32 vs float64 was way smaller. Not 100% proof without a real GPU, but it lines up.

Might be worth trying:

    tf.config.experimental.enable_tensor_float_32_execution(False)

before creating your tensors, see if the gap shrinks a lot. If it does, that'd confirm it's TF32 and not a real precision bug.

(Side note, the Dropout at rate=0.0 in your repro isn't doing anything — it shows 0.0 diff itself — so whatever's happening is all in the matmul.)

Also noticed matmul's docstring never actually mentions TF32, so put up a small docs PR for that either way: #127577.

**Comment by dingjiawei65** [comment]
@RiccardoSquarcialupi Thanks for the explanation. I reran the repro on an RTX 4090 with TensorFlow 2.20 and tested the default setting, TF32 forced on, and TF32 disabled.

With TF32 enabled/default, I get a relative max diff around 2.90e-4, matching the original report. After calling:

tf.config.experimental.enable_tensor_float_32_execution(False)

before creating tensors / running matmul, the relative max diff drops to about 1.45e-6, and the Dropout(rate=0.0) control remains exactly identical across CPU/GPU.

So I agree this is explained by TF32 precision rather than a TensorFlow matmul correctness bug. Thanks for pointing it out. The docs PR mentioning TF32 behavior should help avoid this confusion.
